### PR TITLE
[16.0][MIG] delivery_multi_destination: Migration to version 16.0

### PR DESCRIPTION
**Pull request by carolinafernandez-tecnativa**
- Standard migration to v16
- Supersedes https://github.com/OCA/delivery-carrier/pull/631 due to inactivity/changes requested not attended

@Tecnativa
TT46562

@pedrobaeza @victoralmau

### CODE REVIEW

**Review by chienandalu — CHANGES_REQUESTED** [review]
An issue I found in a functional test:

- I create a multi destination carrier
- With a filtered destination subcarrier
- Then I go to a sale order with a partner no in the filtered destination
- Force choosing that destination in the wizard:

```
Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 1591, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/opt/odoo/odoo/service/model.py", line 133, in retrying
    result = func()
  File "/opt/odoo/odoo/http.py", line 1618, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/opt/odoo/odoo/http.py", line 1822, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "/opt/odoo/odoo/http.py", line 697, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/opt/odoo/addons/web/controllers/dataset.py", line 42, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/addons/web/controllers/dataset.py", line 33, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 468, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 453, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/odoo/models.py", line 6571, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/opt/odoo/odoo/models.py", line 6282, in _onchange_eval
    method_res = method(self)
  File "/opt/odoo/addons/delivery/wizard/choose_delivery_carrier.py", line 32, in _onchange_carrier_id
    vals = self._get_shipment_rate()
  File "/opt/odoo/addons/delivery/wizard/choose_delivery_carrier.py", line 68, in _get_shipment_rate
    if vals.get('success'):
AttributeError: 'NoneType' object has no attribute 'get'

The above server error caused the following client error:
null
```

**Comment by carolinafernandez-tecnativa** [comment]
> An issue I found in a functional test:
> 
> * I create a multi destination carrier
> * With a filtered destination subcarrier
> * Then I go to a sale order with a partner no in the filtered destination
> * Force choosing that destination in the wizard:
> 
> ```
> Traceback (most recent call last):
>   File "/opt/odoo/odoo/http.py", line 1591, in _serve_db
>     return service_model.retrying(self._serve_ir_http, self.env)
>   File "/opt/odoo/odoo/service/model.py", line 133, in retrying
>     result = func()
>   File "/opt/odoo/odoo/http.py", line 1618, in _serve_ir_http
>     response = self.dispatcher.dispatch(rule.endpoint, args)
>   File "/opt/odoo/odoo/http.py", line 1822, in dispatch
>     result = self.request.registry['ir.http']._dispatch(endpoint)
>   File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
>     result = endpoint(**request.params)
>   File "/opt/odoo/odoo/http.py", line 697, in route_wrapper
>     result = endpoint(self, *args, **params_ok)
>   File "/opt/odoo/addons/web/controllers/dataset.py", line 42, in call_kw
>     return self._call_kw(model, method, args, kwargs)
>   File "/opt/odoo/addons/web/controllers/dataset.py", line 33, in _call_kw
>     return call_kw(request.env[model], method, args, kwargs)
>   File "/opt/odoo/odoo/api.py", line 468, in call_kw
>     result = _call_kw_multi(method, model, args, kwargs)
>   File "/opt/odoo/odoo/api.py", line 453, in _call_kw_multi
>     result = method(recs, *args, **kwargs)
>   File "/opt/odoo/odoo/models.py", line 6571, in onchange
>     record._onchange_eval(name, field_onchange[name], result)
>   File "/opt/odoo/odoo/models.py", line 6282, in _onchange_eval
>     method_res = method(self)
>   File "/opt/odoo/addons/delivery/wizard/choose_delivery_carrier.py", line 32, in _onchange_carrier_id
>     vals = self._get_shipment_rate()
>   File "/opt/odoo/addons/delivery/wizard/choose_delivery_carrier.py", line 68, in _get_shipment_rate
>     if vals.get('success'):
> AttributeError: 'NoneType' object has no attribute 'get'
> 
> The above server error caused the following client error:
> null
> ```

I checked this and now it seems to work correctly, could you please check? Thanks!!

File: delivery_multi_destination/models/delivery_carrier.py [hunk 1]
```diff
@@ -69,23 +70,22 @@ def rate_shipment(self, order):
...
-        else:
+        if self.destination_type != "one":
             carrier = self.with_context(show_children_carriers=True)
             for subcarrier in carrier.child_ids:
                 if subcarrier._match_address(order.partner_shipping_id):
                     return super(
                         DeliveryCarrier,
                         subcarrier,
                     ).rate_shipment(order)
+        return super().rate_shipment(order)
```
Comment (chienandalu) [inline delivery_multi_destination/models/delivery_carrier.py]: The error is gone indeed... but this logic looks the same to me :thinking: What am I missing? :sweat_smile:

Comment (pedrobaeza) [reply delivery_multi_destination/models/delivery_carrier.py]: This is not correct. `super` is being called twice.

File: delivery_multi_destination/models/delivery_carrier.py [hunk 2]
```diff
@@ -69,23 +70,22 @@ def rate_shipment(self, order):
...
 
     def send_shipping(self, pickings):
         """We have to override this method for redirecting the result to the
         proper "child" carrier.
         """
         if self.destination_type == "one" or not self:
-            return super().send_shipping(pickings)
+            res = super().send_shipping(pickings)
```
Comment (pedrobaeza) [inline delivery_multi_destination/models/delivery_carrier.py]: Why changing this? It's not needed.

File: delivery_multi_destination/tests/test_delivery_multi_destination.py [hunk 3]
```diff
@@ -84,17 +84,16 @@ def _create_carrier(self, childs):
         carrier_form = Form(self.env["delivery.carrier"])
         carrier_form.name = "Test carrier multi"
         carrier_form.product_id = self.product
-        carrier_form.destination_type = "multi"
         carrier_form.delivery_type = "fixed"
         carrier_form.fixed_price = 100
+        carrier_form.destination_type = "multi"
```
Comment (pedrobaeza) [inline delivery_multi_destination/tests/test_delivery_multi_destination.py:90]: Why changing this order?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: delivery_multi_destination/tests/test_delivery_multi_destination.py [hunk 4]
```diff
@@ -84,9 +84,9 @@ def _create_carrier(self, childs):
         carrier_form = Form(self.env["delivery.carrier"])
         carrier_form.name = "Test carrier multi"
         carrier_form.product_id = self.product
+        carrier_form.destination_type = "multi"
```
Comment (pedrobaeza) [inline delivery_multi_destination/tests/test_delivery_multi_destination.py]: Why changing the order of this line?

File: delivery_multi_destination/models/delivery_carrier.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # Copyright 2016-2020 Tecnativa - Pedro M. Baeza
 # Copyright 2017 Tecnativa - Luis M. Ontalba
 # Copyright 2021 Gianmarco Conte <gconte@dinamicheaziendali.it>
+# Copyright 2024 Tecnativa - Carolina Fernandez
```
Comment (pedrobaeza) [inline delivery_multi_destination/models/delivery_carrier.py]: Remove this copyright

File: delivery_multi_destination/tests/test_delivery_multi_destination.py (hunk 4 above)
Comment (pedrobaeza) [reply delivery_multi_destination/tests/test_delivery_multi_destination.py]: Oh, I see that is moving it back to the original position. Now I see the CI red, so maybe it's impacting in something. If it impacts, put a comment saying that this needs to be done in this order. If not impacting, better to neutralize the change in the previous commit, and don't revert it back in your commit.

Comment (carolinafernandez-tecnativa) [reply delivery_multi_destination/tests/test_delivery_multi_destination.py]: @pedrobaeza done!

**Review by pedrobaeza — APPROVED** [review]
The dependency of the order of the assigned values is a bit weird, but let's not block this more.

/ocabot merge nobump

**Comment by carolinafernandez-tecnativa** [comment]
ping @chienandalu

**Pull request merged**